### ISSUE TITLE
[refactoring] repo-put / repo-get

### ISSUE THREAD

**Issue by pgstef** [pgbackrest/pgbackrest#1090]
Hi,

I was trying to improve the repo-get command encrypted part. 

Until now, here are the encryption levels I found:
* repo cipher -> archive.info & backup.info
* archive cipher, stored in archive.info, applied to WAL archives (and .history)
* backup cipher, stored in backup.info, applied to manifests (and backup.history manifests)
* manifest cipher, stored in backup.manifest, applied to backed-up files

Am I missing something there?

In order to imagine all the possible cases for the repo-get command (like guessing which cipher to use), I think it will be easier to first work on the repo-put command.

ATM, the repo-put command can store pretty much anything with a specific `--cipher-pass`. This cipher won't be stored anywhere. Is that something we wish to keep ?

* make stanza parameter mandatory
* copy file in the according repo archive -> apply archive.info stored cipher (used to push a wal archive eg.)
* copy file in the according repo backup dir -> apply backup.info stored cipher

I'm then considering using `--type` to know where to put files (also meaning that absolute path would only work if it match all the options).

Then, to put a file in a specific backup set, I would use the `--type=backup` + `--set` to apply the manifest specific cipher.

It would give a pretty restricted `repo-put` command, allowing to apply the good cipher.

Does that seems good for you or did you imagine something else ?

**Comment by dwsteele** [comment]
> Until now, here are the encryption levels I found

This looks right to me.

> I think it will be easier to first work on the repo-put command.

Actually, I think `repo-get` is a bit simpler and more useful overall. Also, it's not clear that we would ever make `repo-put` public.

> repo-put command can store pretty much anything with a specific --cipher-pass

This is needed for testing with `repo-put` and `repo-get` but the option will remain hidden from users even if we make the command public.

> I'm then considering using --type ... I would use the --type=backup + --set

This seems much more complicated than necessary and will make the commands very hard to use. All the info needed to is in the path, e.g.
```
backup/test/20200628-084515F/pg_data/backup_label
```
is a backup file which needs to be decrypted with the key in `backup/test/20200628-084515F/backup.manifest` which needs to be decrypted with the key in `backup/test/backup.info` which needs to be decrypted with `repo1-cipher-pass`.

> Does that seems good for you or did you imagine something else ?

In summary, I think the `repo-get` command will be a bit simpler to implement and generally more useful (especially to check_pgbackrest). `repo-get` is also less dangerous, i.e. it can't be used to corrupt the repo.

**Comment by pgstef** [comment]
Hi,
Thanks for your feedback. I'll try to propose a PR in that way.
I close the issue and will reopen it if I've more questions ;-)
Have a nice day
